### PR TITLE
feat: create gov proposal for token metadata registration

### LINKED ISSUE #604: Add gov-proposal for creation/modification/deletion of bank denom metadata

### PR DESCRIPTION
**Pull request by srene**
## Description

This PR creates a new module that will enable creating government proposals for creating or updating token representation metadata

----

Closes #604

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x]  Targeted PR against the correct branch
- [x]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x]  Linked to the GitHub issue with discussion and accepted design
- [x]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: proto/dymension/denommetadata/tokenmetadata.proto [hunk 1]
```diff
@@ -0,0 +1,53 @@
+syntax = "proto3";
+package dymensionxyz.dymension.denommetadata;
```
Comment (omritoptix) [inline proto/dymension/denommetadata/tokenmetadata.proto:2]: why is this file needed? you have this in the cosmos sdk already (proto/cosmos/bank/v1beta1/bank.proto)

File: proto/dymension/denommetadata/genesis.proto [hunk 2]
```diff
@@ -0,0 +1,13 @@
+syntax = "proto3";
+package dymensionxyz.dymension.denommetadata;
+
+import "gogoproto/gogo.proto";
+
```
Comment (omritoptix) [inline proto/dymension/denommetadata/genesis.proto:5]: why is this file needed?

File: x/denommetadata/client/cli/tx.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+		Use:                        types.ModuleName,
+		Short:                      fmt.Sprintf("%s transactions subcommands", types.ModuleName),
+		DisableFlagParsing:         true,
+		SuggestionsMinimumDistance: 2,
+		RunE:                       client.ValidateCmd,
+	}
+
+	cmd.AddCommand(NewCmdSubmitCreateDenomMetadataProposal())
+	cmd.AddCommand(NewCmdSubmitUpdateDenomMetadataProposal())
+	// this line is used by starport scaffolding # 1
```
Comment (omritoptix) [inline x/denommetadata/client/cli/tx.go:29]: redundant comment

File: x/denommetadata/client/cli/utils.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/osmosis-labs/osmosis/v15/osmoutils"
+	"github.com/spf13/cobra"
+)
+
+func parseProposal(cmd *cobra.Command) (osmoutils.Proposal, sdk.Coins, error) {
```
Comment (omritoptix) [inline x/denommetadata/client/cli/utils.go:11]: you already have it in dymension/utils. why creating a new utils for parseProposal?

File: app/app.go [hunk 5]
```diff
@@ -672,7 +683,8 @@ func New(
 		AddRoute(upgradetypes.RouterKey, upgrade.NewSoftwareUpgradeProposalHandler(app.UpgradeKeeper)).
 		AddRoute(ibcclienttypes.RouterKey, ibcclient.NewClientProposalHandler(app.IBCKeeper.ClientKeeper)).
 		AddRoute(streamermoduletypes.RouterKey, streamermodule.NewStreamerProposalHandler(app.StreamerKeeper)).
-		AddRoute(rollappmoduletypes.RouterKey, rollappmodule.NewRollappProposalHandler(&app.RollappKeeper))
+		AddRoute(rollappmoduletypes.RouterKey, rollappmodule.NewRollappProposalHandler(&app.RollappKeeper)).
+		AddRoute(denommetadatamoduletypes.RouterKey, denommetadatamodule.NewDenomMetadataProposalHandler(app.DenomMetadataKeeper))
```
Comment (omritoptix) [inline app/app.go:687]: instead of creating a new DenomMetaDAtaKEeper which doesn't do too much why not just pass here the app.BankKeeper and do that logic in the handlers and remove the denommetaata keeper all together?

File: x/denommetadata/module.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
+package denommetadata
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/gorilla/mux"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/spf13/cobra"
+	abci "github.com/tendermint/tendermint/abci/types"
```
Comment (omritoptix) [inline x/denommetadata/module.go:10]: not sure this file is needed if you get rid of the denommetadat keeper like I suggested in previosu comment

File: x/denommetadata/types/keys.go [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package types
+
+var (
+	// ModuleName defines the module name.
+	ModuleName = "denommetadata"
+
+	// StoreKey defines the primary module store key.
+	StoreKey = ModuleName
+
+	// RouterKey is the message route for slashing.
```
Comment (omritoptix) [inline x/denommetadata/types/keys.go:10]: wrong godoc

File: x/denommetadata/types/genesis.go [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package types
+
+// this line is used by starport scaffolding # genesis/types/import
+
+// DefaultIndex is the default global index
+const DefaultIndex uint64 = 1
```
Comment (omritoptix) [inline x/denommetadata/types/genesis.go:6]: redundant file

File: x/denommetadata/types/tonemmetadata.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
+package types
+
+import (
+	"fmt"
+
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+func NewTokenMetadata(description string, denomstring string, exponent uint32, denomalias string, base string, display string, name string, symbol string, uri string, urihash string) *TokenMetadata {
```
Comment (omritoptix) [inline x/denommetadata/types/tonemmetadata.go:9]: redundant file if u use the bank denommetadata type

File: x/denommetadata/keeper/denommetadata.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/gogo/protobuf/proto"
+)
+
+// GetDenomMetadataByID returns denommetadata from denommetadata ID.
+func (k Keeper) GetDenomMetadataByID(ctx sdk.Context, denomMetadataID uint64) (*types.DenomMetadata, error) {
+	denommetadata := types.DenomMetadata{}
+	store := ctx.KVStore(k.storeKey)
+	denomMetadataKey := denomMetadataStoreIdKey(denomMetadataID)
+	if !store.Has(denomMetadataKey) {
+		return nil, fmt.Errorf("DenomMetadata with ID %d does not exist", denomMetadataID)
```
[the remark this answers is not in the document]
Comment (srene) [reply x/denommetadata/keeper/denommetadata.go]: fixed

File: x/denommetadata/client/cli/tx.go (hunk 3 above)
Comment (srene) [reply x/denommetadata/client/cli/tx.go:29]: removed

File: proto/dymension/denommetadata/tokenmetadata.proto (hunk 1 above)
Comment (srene) [reply proto/dymension/denommetadata/tokenmetadata.proto]: removed

File: x/denommetadata/client/cli/utils.go (hunk 4 above)
Comment (srene) [reply x/denommetadata/client/cli/utils.go]: removed

File: app/app.go (hunk 5 above)
Comment (srene) [reply app/app.go]: done

File: x/denommetadata/types/proposal_metadata.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+func (csp *CreateDenomMetadataProposal) GetDescription() string { return csp.Description }
+
+// GetDescription returns the routing key of a community pool spend proposal.
+func (csp *CreateDenomMetadataProposal) ProposalRoute() string { return RouterKey }
+
+// ProposalType returns the type of a community pool spend proposal.
+func (csp *CreateDenomMetadataProposal) ProposalType() string { return ProposalTypeCreateDenomMetadata }
+
+// ValidateBasic runs basic stateless validity checks
+func (csp *CreateDenomMetadataProposal) ValidateBasic() error {
```
Comment (mtsitrin) [inline x/denommetadata/types/proposal_metadata.go:51]: I think the Metadata needs to be validated here as well

File: x/denommetadata/module.go (hunk 6 above)
Comment (srene) [reply x/denommetadata/module.go:10]: without a module if fails to resolve the URL for the tx

File: x/denommetadata/types/keys.go (hunk 7 above)
Comment (srene) [reply x/denommetadata/types/keys.go:10]: fixed

File: x/denommetadata/types/tonemmetadata.go (hunk 9 above)
Comment (srene) [reply x/denommetadata/types/tonemmetadata.go]: fixed

File: x/denommetadata/types/proposal_metadata.go (hunk 11 above)
Comment (srene) [reply x/denommetadata/types/proposal_metadata.go:51]: validation added

File: x/denommetadata/types/genesis.go (hunk 8 above)
Comment (srene) [reply x/denommetadata/types/genesis.go:6]: needed for the module

**Review by omritoptix — APPROVED** [review]
